File: python/tutorial2_control_camera.py
import numpy as np
import cv2

import os
os.add_dll_directory(os.path.join(os.environ["SENSING_DEV_ROOT"], "bin"))

from ionpy import Node, Builder, Buffer, PortMap, Port, Param, Type, TypeCode

if __name__ == "__main__":

    # device information
    width = 1920
    height = 1080
    pixelformat = "Mono8"
    num_device = 1

    # the following items varies by PixelFormat
    data_type = np.uint8 if pixelformat == "Mono8" or pixelformat == "RGB8" \
        else np.uint16 if pixelformat == "Mono10" or pixelformat == "Mono12" \
        else np.uint8
    depth_of_buffer = np.iinfo(data_type).bits
    depth_in_byte = int(depth_of_buffer / 8)
    num_bit_shift = 0 if pixelformat == "Mono8" or pixelformat == "RGB8" \
        else 4 if pixelformat == "Mono12" \
        else 6 if pixelformat == "Mono10" \
        else 0
    bb_name = "image_io_u3v_cameraN_u8x2" if pixelformat == "Mono8" \
        else "image_io_u3v_cameraN_u16x2" if pixelformat == "Mono10" or pixelformat == "Mono12" \
        else "image_io_u3v_cameraN_u8x3" if pixelformat == "RGB8" \
        else "image_io_u3v_cameraN_u8x2"
    
    # pipeline setup
    builder = Builder()
    builder.set_target('host')
    builder.with_bb_module('ion-bb')

    gain_ps = []
    exposure_ps = []
    for i in range(num_device):
        gain_ps.append(Port('gain' + str(i), Type(TypeCode.Float, 64, 1), 0))
        exposure_ps.append(Port('exposure' + str(i), Type(TypeCode.Float, 64, 1), 0))

    # set params
    num_devices = Param('num_devices', str(num_device))
    frame_sync = Param('frame_sync', 'false')
    realtime_display_mode = Param('realtime_display_mode', 'true')
    enable_control = Param('enable_control', 'true')
    gain_key = Param('gain_key', 'Gain')
    exposure_key = Param('exposure_key', 'ExposureTime')

    gain_values = []
    exposure_values = []

    for i in range(num_device):
        gain_values.append(40.0)
        exposure_values.append(100.0)

    # add a node to pipeline
    node = builder.add(bb_name)\
        .set_iport([gain_ps[0], exposure_ps[0]])\
        .set_param([num_devices, frame_sync, realtime_display_mode, enable_control, gain_key, exposure_key]) if num_device == 1 \
        else builder.add(bb_name)\
            .set_iport([gain_ps[0], exposure_ps[0], gain_ps[1], exposure_ps[1]])\
            .set_param([num_devices, frame_sync, realtime_display_mode, enable_control, gain_key, exposure_key])
    output_p = node.get_port('output')

    # create halide buffer for output port
    outputs = []
    output_datas = []
    output_size = (height, width, )
    if pixelformat == "RGB8":
        output_size += (3,)
    for i in range(num_device):
        output_datas.append(np.full(output_size, fill_value=0, dtype=data_type))
        outputs.append(Buffer(array= output_datas[i]))

    # set I/O ports
    for i in range(num_device):
        gain_ps[i].bind(gain_values[i])
        exposure_ps[i].bind(exposure_values[i])
        output_p[i].bind(outputs[i])

    # prepare Opencv 
    buf_size_opencv = (height, width)
    output_byte_size = width*height*depth_in_byte
    if pixelformat == "RGB8":
        buf_size_opencv += (3,)
        output_byte_size *= 3

    coef = pow(2, num_bit_shift)
    user_input = -1

    while(user_input == -1):
        # running the builder
        builder.run()
        
        for i in range(num_device):
            output_datas[i] *= coef

            cv2.imshow("img" + str(i), output_datas[i])
        user_input = cv2.waitKeyEx(1)

    cv2.destroyAllWindows()

File: cpp/src/tutorial2_control_camera.cpp
/*

g++ src/tutorial2_control_camera.cpp -o tutorial2_control_camera  \
-I /opt/sensing-dev/include -I /opt/sensing-dev/include/aravis-0.8 \
-I /opt/sensing-dev/include/opencv4 \
-L /opt/sensing-dev/lib \
-L /opt/sensing-dev/lib/x86_64-linux-gnu \
-lHalide -lion-core -ldl -lpthread -lopencv_core \
-lopencv_imgcodecs -lopencv_highgui -lopencv_imgproc \
-laravis-0.8 -lgobject-2.0 \
`pkg-config --cflags --libs glib-2.0`

*/

#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

#include <ion/ion.h>

#include <exception>
#include <string>
#include <map>

using namespace ion;

std::map<std::string, int32_t> num_bit_shift_map{
  {"Mono8", 0}, {"Mono10", 6}, {"Mono12", 4}};
std::map<std::string, int32_t> opencv_mat_type{
  {"Mono8", CV_8UC1}, {"Mono10", CV_16UC1}, {"Mono12", CV_16UC1}};
std::map<std::string, std::string> bb_name{
  {"Mono8", "image_io_u3v_cameraN_u8x2"}, 
  {"Mono10", "image_io_u3v_cameraN_u16x2"}, 
  {"Mono12", "image_io_u3v_cameraN_u16x2"}};

int positive_pow(int base, int expo){
  if (expo <= 0){
      return 1;
  }
  if (expo == 1){
      return base;
  }else{
      return base * positive_pow(base, expo-1);
  }
}

double gain0 = 35.0;
double exposuretime0 = 50.0;
double gain1 = 47.0;
double exposuretime1 = 100.0;

template<typename T>
int video(int width, int height, std::string pixel_format, int num_device){
    // pipeline setup
    Builder b;
    b.set_target(ion::get_host_target());
    b.with_bb_module("ion-bb");

    // add BB to pipeline
    Node n;
    if (num_device == 1){
      n = b.add(bb_name[pixel_format])(&gain0, &exposuretime0)
        .set_params(
          Param("num_devices", num_device),
          Param("frame_sync", true),
          Param("realtime_display_mode", true),
          Param("enable_control", true),
          Param("gain_key", "Gain"),
          Param("exposure_key", "ExposureTime")
        );
    }else if (num_device == 2){
      n = b.add(bb_name[pixel_format])(&gain0, &exposuretime0, &gain1, &exposuretime1)
        .set_params(
          Param("num_devices", num_device),
          Param("frame_sync", true),
          Param("realtime_display_mode", true),
          Param("enable_control", true),
          Param("gain_key", "Gain"),
          Param("exposure_key", "ExposureTime")
        );
    }else{
      throw("Set the same number of gains and exposure times as the number of devices.");
    }
    
    // portmapping from output port to buffer
    std::vector< int > buf_size = std::vector < int >{ width, height };
    if (pixel_format == "RGB8"){
        buf_size.push_back(3);
    }
    std::vector<Halide::Buffer<T>> output;
    for (int i = 0; i < num_device; ++i){
      output.push_back(Halide::Buffer<T>(buf_size));
    }
    n["output"].bind(output);

    int coef =  positive_pow(2, num_bit_shift_map[pixel_format]);
    int user_input = -1;

    while(user_input == -1)
    {
      // JIT compilation and execution of pipelines with Builder.
      b.run();
      
      // Convert the retrieved buffer object to OpenCV buffer format.
      for (int i = 0; i < num_device; ++i){
        cv::Mat img(height, width, opencv_mat_type[pixel_format], output[i].data());
        img *= coef;
        cv::imshow("image" + std::to_string(i), img);
      }

      // Wait for 1ms
      user_input = cv::waitKeyEx(1);
    }
    return 0;
}

int main(int argc, char *argv[])
{
  try{
    int32_t width = 1920;
    int32_t height = 1080;
    int32_t num_device = 1;
    std::string pixelformat = "Mono8";

    if (pixelformat == "Mono8"){
      int ret = video<uint8_t>(width, height, pixelformat, num_device);
    }else{
      int ret = video<uint16_t>(width, height, pixelformat, num_device);
    }

  } catch(std::exception& e){
    std::cerr << e.what() << std::endl;
    return 1;
  } catch (const ion::Error& e) {
    std::cerr << e.what() << std::endl;
    return 1;
  }
  return 0;
}